File: backend/apps/plugins/registry.py
class PluginRegistry:
    _plugins = {}

    @classmethod
    def register(cls, plugin_class):
        instance = plugin_class()
        cls._plugins[instance.plugin_type] = instance
        return plugin_class

    @classmethod
    def get_plugin(cls, plugin_type):
        return cls._plugins.get(plugin_type)

    @classmethod
    def get_active_plugins(cls):
        from .models import Plugin
        active_types = Plugin.objects.filter(is_active=True).values_list("type", flat=True)
        return {ptype: cls._plugins[ptype] for ptype in active_types if ptype in cls._plugins}

    @classmethod
    def get_all_plugins(cls):
        return dict(cls._plugins)

    @classmethod
    def get_profile_plugins(cls, user_id: int) -> list:
        result = []
        for plugin in cls.get_active_plugins().values():
            data = plugin.get_profile_data(user_id)
            if data:
                result.append(data)
        return result


def register_plugin(cls):
    PluginRegistry.register(cls)
    return cls

File: frontend/src/plugins/registry.js
class PluginRegistry {
  constructor() {
    this.plugins = new Map()
  }

  register(pluginConfig) {
    this.plugins.set(pluginConfig.type, pluginConfig)
  }

  getPlugin(type) {
    return this.plugins.get(type)
  }

  getActivePlugins() {
    return Array.from(this.plugins.values())
  }

  getRoutes() {
    const routes = []
    for (const plugin of this.plugins.values()) {
      if (plugin.route) {
        routes.push(plugin.route)
      }
    }
    return routes
  }
}

export const pluginRegistry = new PluginRegistry()
